src/devtools/client/debugger/src/components/SecondaryPanes/Frames/NewFrames.tsx

```diff
@@ -0,0 +1,163 @@
+import {
+  getFrameworkGroupingState,
+  getPauseId,
+  getSelectedFrame,
+  getThreadContext,
+} from "devtools/client/debugger/src/selectors";
+import { getFramesSuspense } from "ui/suspense/frameCache";
+import { collapseFrames, formatCopyName } from "../../../utils/pause/frames";
+import { createFrame } from "../../../client/create";
+import { ThreadFront } from "protocol/thread";
+import Frame from "./Frame";
+import Group from "./Group";
+import { CommonFrameComponentProps } from ".";
+import { toggleFrameworkGrouping as setFrameworkGroupingAction } from "../../../reducers/ui";
+import { selectFrame as selectFrameAction } from "../../../actions/pause/selectFrame";
+import { copyToTheClipboard } from "../../../utils/clipboard";
+import { PauseId } from "@replayio/protocol";
+import { getSourceDetailsEntities, getSourcesLoading } from "ui/reducers/sources";
+import { formatCallStackFrames } from "../../../selectors/getCallStackFrames";
+import { useAsyncParentPauseIdSuspense } from "ui/suspense/util";
+import ErrorBoundary from "bvaughn-architecture-demo/components/ErrorBoundary";
+import { isCurrentTimeInLoadedRegion } from "ui/reducers/app";
+import { enterFocusMode as enterFocusModeAction } from "ui/actions/timeline";
+
+function FramesRenderer({
+  panel,
+  pauseId,
+  asyncIndex = 0,
+}: {
+  panel: "debugger" | "console" | "webconsole";
+  pauseId: PauseId;
+  asyncIndex?: number;
+}) {
+  const state = useAppSelector(state => state);
+  const cx = getThreadContext(state);
+  const sources = getSourceDetailsEntities(state);
+  const frameworkGroupingOn = getFrameworkGroupingState(state);
+  const selectedFrame = getSelectedFrame(state);
+  const dispatch = useAppDispatch();
+
+  function toggleFrameworkGrouping() {
+    dispatch(setFrameworkGroupingAction(!frameworkGroupingOn));
+  }
+  function selectFrame(...args: Parameters<typeof selectFrameAction>) {
+    dispatch(selectFrameAction(...args));
+  }
+  //TODO copy async parent frames as well
+  function copyStackTrace() {
+    const framesToCopy = frames?.map(f => formatCopyName(f)).join("\n");
+    copyToTheClipboard(framesToCopy);
+  }
+
+  const commonProps: CommonFrameComponentProps = {
+    cx,
+    toggleFrameworkGrouping,
+    frameworkGroupingOn,
+    selectFrame,
+    selectedFrame,
+    displayFullUrl: false,
+    disableContextMenu: false,
+    copyStackTrace,
+    panel,
+  };
+
+  let protocolFrames = getFramesSuspense(pauseId);
+  if (asyncIndex > 0) {
+    protocolFrames = protocolFrames?.slice(1);
+  }
+  const frames = useMemo(
+    () =>
+      formatCallStackFrames(
+        protocolFrames?.map((protocolFrame, index) =>
+          createFrame(
+            state,
+            ThreadFront.preferredGeneratedSources,
+            protocolFrame,
+            index,
+            asyncIndex
+          )
+        ) || null,
+        sources
+      ),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
:question: what dep are we avoiding a warning for here?

Reply:
This feels like a foot gun. If we use it to look at sources (or anything) then it's a dependency, isn't it?

Can we pre-extract sources and use _them_ as the dependency then?

Generally, bypassing the exhaustive deps rule seems like a bad idea to me.

Reply:
Inclined to agree with Brian on principle here. It's worth rearranging function signatures so we don't have to do this and can instead just pass in the data that's needed.